### DISCUSSION TITLE
Using Fonts in GOODF

### DISCUSSION THREAD

**Discussion by davidgritter** [GrandOrgue/GoOdf#13] [General]
What are the limitations for fonts in GOODF/GrandOrgue.  When using GOODF I can select a font, say **Noto Sans ExtraCondensed Bold**, point size 10.   However when written into the ODF it just appears as Noto Sans, font size LARGE. I can change the point size just fine in the ODF and get the proper size, but entering the full font name in bold above seems to do nothing,  Is this a limitation of GrandOrgue that GOODF knows about?

**Comment by larspalo** [comment]
I'm not completely sure about this. However, both GOODF and GO uses wxWidgets for handling the GUI in similar ways. It could be that the wxFont::GetFaceName() returns just the font name without the extra information (ExtraCondensedBold), and GO uses ultimately wxFont::SetFaceName() for creating the font which might do the same.

**Reply by davidgritter** [reply]
Yes I see.  There are different functions to completely specify a font, but if this info needs to be stored as text, as in a  .organ file, it seems these functions are different for different platforms.

**Comment by davidgritter** [comment]
I have several suggestions to improve the workflow with fonts in GOODF, for you to consider at your discretion.
1.  default fonts are chosen in the Display Metrics window.  This includes both a font name and a point size.  however the [panel] section of the ODF doesn't allow the font point size to be included.  Consider automatically carrying this point size down to elements that don't define their own font.   You could also consider selecting a default color in the display metrics window to be carried down.
2. Consider writing the actual point size of the font to the ODF for panel elements instead of the small, medium, Large.  I have tried to define point sized up to 20 points for some elements, but the size in the odf remains at large, which gives something around 10-12 points.

**Comment by larspalo** [comment]
So for your first point, you want the display metrics control label font to become the default value of all button types on that panel. Also the group label font should become the default for all labels on that panel. That makes sense as that's how GO would handle things. To include the size is not a problem, except for that on opening an .organ file there's no value to get the default from. But it could work within GOODF as long as one is working on the odf.

If the size is not written in points if the value is something else than 6 (small), 7 (normal) or 10 (large) there's a bug somewhere that needs fixing.

I'll investigate it and fix the default font settings for button types and labels.

**Comment by larspalo** [comment]
I've committed fixes for all button types, labels and also enclosure (though there the defaults are not derived from display metrics). Please test it.

For the colors, I think that the base defaults of GO will be what I'll also keep for each element in GOODF.

**Comment by davidgritter** [comment]
In testing this commit I note that font size for the elements is written in points now, as it should be.

However, for each GUI element (switch is what I tested) GOODF now writes a line  "DispLabelFontName=Arial" if the font for that element shows up as "None" on its GOODF window (the font defined in dispaly.  This behavior is different than previous versions, which correctly wrote no DispLabelFontName line so that the font specified in the displaymetrics was used.

I would have expected instead that if the Gui element had a font of "None", that GOODF would write a "DispLabelFontSize=12" if the point size in displaymetrics was set to 12.  This is particularly important since one can't specify a size without also specifying a font in the GUIelement GOODF window.
Thanks for your consideration

**Reply by larspalo** [reply]
The intended behavior is to write out the font name for a button derived type only if it differs from the display metrics dispControlLabelFont. The font size should be handled separately, which already "should" work inside GOODF, but I think I'll need to fix the transfer of size into the font when an .organ file is opened because that seems to not be working at the moment...

I'm actually more worried about the fact that you can get a font of "None". That in itself is something that needs investigation as I've never seen that happen myself so far.

What I tried to implement is that at element creation (even before the element could be read from odf) I set the default to the panel display metrics font. If the read value then differs it will override the default. I didn't check what would happen if the font of an element was empty though...

**Reply by larspalo** [reply]
There is a new commit that actually applies the font size of a read element to the font itself for that element. This "should" work even if no font is defined for the element and the size then is applied to the default font derived from the display metrics.

The default size from display metrics inside GOODF should transfer to any new GUI element that's created, but this of course won't work with already created elements, nor when read from odf as GO has another opinion of what the default font size should be for the elements than the display metrics (non existing font size).

**Reply by davidgritter** [reply]
"I'm actually more worried about the fact that you can get a font of "None". That in itself is something that needs investigation as I've never seen that happen myself so far."    

I don't get a font of "none" in the ODF, only in the GOODF element window if a font is not yet defined.  What happened in the testing above, is that GOODF wrote out a FontName of "arial" for those elements that had "none" in GOODF, instead of not writing a FontName line at all

**Reply by larspalo** [reply]
> I don't get a font of "none" in the ODF, only in the GOODF element window if a font is not yet defined. What happened in the testing above, is that GOODF wrote out a FontName of "arial" for those elements that had "none" in GOODF, instead of not writing a FontName line at all

The only reason for GOODF now to write out the font name to a button type, or label, element is if the font of the element doesn't match the display metrics control/group label font. If they match, no font will (should) be written to the element. 

GOODF still provide Arial as a default font when a new organ is created in GOODF both for the display metrics and all elements. This could perhaps be causing problems as Arial is not a "standard" font that exist in Linux distributions. I hope that the font is translated to something similar, but usable, by wxWidgets. If this indeed is a problem, I think that one could default to just a font family instead inside GOODF when creating the defaults. But I think that on all my Linux machines I have at least the basic windows fonts present, so Arial will always be a valid font here when I test it.

My later commits should provide the elements the same font by default as the display metrics if the element doesn't specify one themselves when they are read. This font should also be visible in the element gui panels inside GOODF. I also hope that now the font size is read and applied properly.

**Comment by davidgritter** [comment]
1: Arial does not cause a problem even if Arial font is not installed

2: I have tested your latest commits and agree that the initial write of the ODF is correct and as anticipated.  However, I propose, in the attached document, a USE CASE for GOODF that the present implementation cannot accomodate.  This USE CASE is the situation where the user wants to load an existing ODF created by GOODF and change the font characteristics for multiple GUI elements simply by editing the font information in the displaymetrics window.
[font inheritance.odt](https://github.com/larspalo/GOODF/files/11191505/font.inheritance.odt)

**Reply by larspalo** [reply]
@davidgritter For your second point: What if we would add an option on the display metrics to apply selected font name (not size as that cannot be trusted from the display metrics) to all existing elements in that panel (that derive from it)? And we could as well add an option on each gui element to revert back to display metrics derivation of the font? This could be buttons that just require a single click to apply, but it's also possible to just add a message dialog that will fire after the font selection dialog that ask if this font should apply to all gui elements (for the display metrics), or for a single gui element a cancel of the font selection dialog could trigger a message dialog with a similar question if the user wants to revert to the default display metrics font name. A reason for the second type is that it won't take up any more space in the panels!

Such a solution would make it possible to quickly revert back for all elements as well as (fairly) quickly do that for individual elements without searching for what font should be set. Opinions on that?

The simple fact that gui element values (always should) override the display metrics values in odf stipulate that such a feature will require manual input so that it must be a conscious choice to apply display metrics font name to all existing elements - as this could very well not be a desired outcome too...

**Reply by davidgritter** [reply]
any of these approaches you suggest should meet the objectives of my USE CASE, and are probably more compatible with how the rest of GOODF works. I would argue, however, that any button or message dialog on a GUI element apply only the displaymetrics font as you suggest, but that on any option on the displaymetrics should apply both the font name and (perhaps optionally) the font size, since a typical USE CASE is that the user only wants to change the font size on all his GUIelements.

**Reply by larspalo** [reply]
@davidgritter I definitely think that the font size and font name should be possible to apply separately from each other to multiple elements, maybe a two step confirmation/cancel could solve that.

After changing font on the display metrics panel (or opening the font picker dialog) upon OK one could imagine asking the user if he wants to apply the selected font name to all GUI elements (do so if OK or don't if NO is chosen) and after that ask the user if he wants to apply selected font size to all GUI elements.

Should I try to implement it like that?

**Reply by larspalo** [reply]
In latest commit I've made it possible to "batch change" font names and/or size.

**Reply by larspalo** [reply]
@davidgritter When you've tested this, and if you think it's working, please also review the currently open issues so that none of them still is active. If they all are fixed, I think it's about time to create a new release since a lot of bugs have been fixed since the last one.

**Reply by davidgritter** [reply]
This seems to work as expected.  The only downside I see is that the font will be applied to all button objects, but generals typically have a larger font size than drawstops.  But there is probably no easy solution for this, so we have to live with it. On most of my own ODFs generals (and divisionals if present) are on a different panel anyway.

I have attached an updated help file describing these features
[creatinggui.tar.gz](https://github.com/larspalo/GOODF/files/11195487/creatinggui.tar.gz)

**Comment by davidgritter** [comment]
Issues 10 and 11 are resolved and can be closed in my opinion.  Issue 12 w/regard to spacing and comments I have not tested, but I have no  arguement against your proposed solution.  We could possibly put a note about this in the help for the few users who manually edit their ODFs.
